devito/at_controller.py

```diff
@@ -0,0 +1,344 @@
+        shutil.copyfile(at_report_path, os.path.join(reports_folder_path, "%s-at-report.txt" % report_name_ref))
+
+    def _extract_best_block_size(self, at_report_path):
+        """Extracts the best best block size from at report and writes it to  the final report file
+
+        Args:
+            at_report_path (str): full path to a at report
+
+        Raises:
+            ValueError: If failed to extract block size. Can't open report files, can't write to them. etc
+        """
+        global final_report_path
+        try:
+            # Gets the required info from  at report
+            with open(at_report_path, 'r') as at_report:
+                # if all these keywords are in the line then next line is the one we want
+                keyword_lst = ["Rank", "Value", "Time(secs)"]
+
+                lines = at_report.readlines()
+                for i in range(0, len(lines)):  # loops through all the lines trying to find the best
+                    if all(keyword in lines[i] for keyword in keyword_lst):
+                        #                                   finds text between parenthesis
+                        best_block = lines[i + 1][lines[i + 1].find("(") + 1:lines[i + 1].find(")")]
+                        best_block = best_block.replace(" ", '')
+
+                        # writes the best block size for dimensions starting from outer most
+                        str_to_write = "\n%d %d %s" % (self.time_order, self.space_order, best_block)
+                        break
+
+            if not os.path.isfile(final_report_path):  # initialises report file if it does not exist
+                with open(final_report_path, 'w') as final_report:
+                    final_report.write("time o,space o,best block size")
+                    final_report.write(str_to_write)  # writes the string
+            else:
+                with open(final_report_path, 'r') as final_report:  # reads all the contents
+                    lines = final_report.readlines()
+
+                # checks whether entry already exist and updates it. Otherwise appends to the end of the file
+                entry_found = False
+                str_to_check = "%d %d " % (self.time_order, self.space_order)
+                for i in range(1, len(lines)):
+                    if str_to_check in lines[i]:  # remove the newline from beginning of the string
+                        lines[i] = str_to_write.replace('\n', '')  # replace line if matching found
+                        entry_found = True
+                        break
+
+                if not entry_found:  # if entry not found append string to the end of file
+                    lines.append(str_to_write)
+
+                with open(final_report_path, 'w') as final_report:  # writes all the contents
+                    final_report.writelines(lines)
+
+        except Exception:
+            print "Failed to extract best block size from %s" % at_report_path
+            raise ValueError()
```

Comment:
Maybe not, however I dont think it matters as parent method  should catch all the exceptions to prevent AT process from intervening with actual code.